http: add StringMatcher in HeaderMatcher and deprecate the old fields (exact, prefix, etc.)

yangminzhu opened the pull request:
Signed-off-by: Yangmin Zhu <ymzhu@google.com>

<!--
!!!ATTENTION!!!

If you are fixing *any* crash or *any* potential security issue, *do not*
open a pull request in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.
Thank you in advance for helping to keep Envoy secure.

!!!ATTENTION!!!

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)
-->

Commit Message: Add StringMatcher in HeaderMatcher and deprecate the old fields (exact, prefix, etc.)
Additional Description:
Risk Level: Low
Testing: Unit Tests
Docs Changes:
Release Notes: Added notes for the new field `string_match` and the deprecation of the old fields (exact, prefix, etc.).
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

yangminzhu commented:
Could someone take a look at this? The `envoy-presubmit` failed seems not related to the PR, thanks.

soulxu commented on test/common/http/header_utility_test.cc:
```diff
@@ -414,6 +416,25 @@ invert_match: true
 
+TEST(MatchHeadersTest, HeaderExactMatchIgnoreCase) {
+  TestRequestHeaderMapImpl matching_headers_1{{"match-header", "match-value"}};
+  TestRequestHeaderMapImpl matching_headers_2{{"match-header", "MATCH-VALUE"}};
```
also can test a case like 'Match-Value'

yangminzhu replied:
done.

soulxu commented on test/common/http/header_utility_test.cc:
```diff
@@ -414,6 +416,25 @@ invert_match: true
+                                              {"other-header", "match-value"}};
+  const std::string yaml = R"EOF(
+name: match-header
+exact_match: match-value
```
Also can have a test case for 'Match-Value'

yangminzhu replied:
done.

soulxu reviewed (commented):
@yangminzhu the `envoy-presubmit` doesn't seems your fault. :)

yangminzhu commented on source/common/http/header_utility.cc:
```diff
@@ -157,13 +164,18 @@ bool HeaderUtility::matchHeaders(const HeaderMap& request_headers, const HeaderD
-    match = absl::StrContains(header_value.result().value(), header_data.value_);
+    match = header_data.ignore_case_
+                ? absl::StrContains(absl::AsciiStrToLower(value),
+                                    header_data.contains_match_lowercase_.get())
```
It's to cache the lower case result to make it a bit faster so that we do not call AsciiStrToLower() on every requests.

htuch replied:
Why not just use the `value_` field for this?

yangminzhu replied:
changed to StringMatcher now, this is no longer needed.

htuch commented on api/envoy/config/route/v3/route_components.proto:
```diff
@@ -1929,6 +1929,10 @@ message HeaderMatcher {
   // * The range [-10,0) will match the value -1, so it will not match when inverted.
   bool invert_match = 8;
+
+  // If true, indicates the exact_match/prefix_match/suffix_match/contains_match should be case
```
Nit `*exact_match*`, `*safe_regex*` etc.

yangminzhu replied:
updated to use `StringMatcher string_match` instead.

htuch commented on test/common/http/header_utility_test.cc:
```diff
@@ -671,6 +751,30 @@ invert_match: true
 }
 
+TEST(MatchHeadersTest, HeaderContainsMatchIgnoreCase) {
+  for (const auto& match_value : {"onevalue", "OneValue"}) {
```
Why is this a `for` loop? Here and other cases.

yangminzhu replied:
not needed now.

htuch reviewed (changes requested):
Thanks, seems reasonable, just some minor comments.
/wait

markdroth commented on api/envoy/config/route/v3/route_components.proto:
```diff
@@ -1929,6 +1929,10 @@ message HeaderMatcher {
   bool invert_match = 8;
+
+  // If true, indicates the exact_match/prefix_match/suffix_match/contains_match should be case
+  // insensitive. This has no effect for the safe_regex_match.
```
Why does this not apply to `safe_regex_match` as well?

yangminzhu replied:
I'm following the same [ignore_case](https://github.com/envoyproxy/envoy/blob/dfef049fe9692ff8188a83bab859b310c212ece1/api/envoy/type/matcher/v3/string.proto#L65) in StringMatcher which also does not apply to regex, I think probably the reason is you can always enable the case-insensitive match easily with the `(?i)` flag directly in the regular expression itself?

markdroth replied:
Oh, interesting -- I didn't realize that.  And it looks like gRPC C-core has the wrong behavior here.  I'll fix that separately.

yangminzhu replied:
If you think it would be useful, I think we can support it in StringMatcher in a follow PR later.

markdroth replied:
No, that's okay.  I've already fixed gRPC C-core to match what Envoy does (grpc/grpc#26567).  The other gRPC languages do the same thing.

markdroth commented on api/envoy/config/route/v3/route_components.proto:
```diff
@@ -1929,6 +1929,10 @@ message HeaderMatcher {
+
+  // If true, indicates the exact_match/prefix_match/suffix_match/contains_match should be case
+  // insensitive. This has no effect for the safe_regex_match.
+  bool ignore_case = 13;
```
Instead of adding this here, I think it would make more sense to add a [`type.matcher.v3.StringMatcher`](https://github.com/envoyproxy/envoy/blob/dfef049fe9692ff8188a83bab859b310c212ece1/api/envoy/type/matcher/v3/string.proto#L20) field and deprecate the existing `exact_match`, `safe_regex_match`, `prefix_match`, `suffix_match`, and `contains_match` fields.  That way, we can eventually get away from having duplicate implementations here.

Note that `StringMatcher` already supports an `ignore_case` field.

yangminzhu replied:
Sure I'm fine with StringMatcher, it is a slightly larger API change and deprecation. Also cc @htuch in case if there are any other comments about this.

htuch replied:
I think we discussed this in the past and there was a concern we might break a lot of exiting users via a deprecation here. OTOH, since we don't have major versions coming up, I think we need some way to pay down API tech debt. I'm good with this if other API shepherds agree, but would like to check with @mattklein123 who had the reservation last time this came up.

mattklein123 replied:
Yeah I agree this is going to be a very large change for everyone, and kind of annoying, but as @htuch says if v4 is "never" we need to go ahead and do stuff like this. Given that now we are talking about 2 year+ cycles for deprecation removal I think it's fine to go ahead and do it.

yangminzhu replied:
@htuch @mattklein123 thanks for the comment, the PR is ready for review, PTAL.
I updated again and all tests are passing except the Mac CI test timed out after 3 hours which seems not related to the PR.

yangminzhu commented:
@htuch @markdroth updated to use StringMatcher now, PTAL, thanks.

yangminzhu commented on api/envoy/type/matcher/v3/string.proto:
```diff
@@ -62,8 +62,8 @@ message StringMatcher {
 
-  // If true, indicates the exact/prefix/suffix matching should be case insensitive. This has no
-  // effect for the safe_regex match.
+  // If true, indicates the exact/prefix/suffix/contains matching should be case insensitive. This
```
note: this is already supported but just not documented in the API.

markdroth reviewed (commented):
This approach looks great to me!  I'll let Harvey do the API approval.

soulxu commented on configs/terminate_http2_post.yaml:
```diff
@@ -32,7 +32,8 @@ static_resources:
                   headers:
                   - name: ":method"
-                    exact_match: "POST"
+                    string_match:
```
nit, doing a search in the code, still found some example need to update  for example:
https://github.com/envoyproxy/envoy/blob/main/docs/root/configuration/http/http_filters/tap_filter.rst
https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/rbac/v3/rbac.proto#L63

yangminzhu replied:
I was planning to only change those causing the test failure but it seems we'd better just change all of them.

yangminzhu commented:
@htuch would you mind to take a look at the PR again? I changed to use StringMatcher directly suggested by Mark and updated existing config and code to use it accordingly, thanks.

htuch commented on test/tools/router_check/router.cc:
```diff
@@ -463,12 +486,16 @@ bool RouterCheckTool::matchHeaderField(const HeaderMap& header_map,
+  case envoy::config::route::v3::HeaderMatcher::HeaderMatchSpecifierCase::kStringMatch:
+    if (header.string_match().match_pattern_case() !=
+        envoy::type::matcher::v3::StringMatcher::MatchPatternCase::kExact) {
+      goto default_case;
```
Nit: not a huge fan of `goto` in C++

yangminzhu replied:
Updated to remove `FALLTHRU` instead.

htuch commented on test/common/http/header_utility_test.cc:
```diff
@@ -289,6 +289,21 @@ contains_match: somevalueinside
   EXPECT_EQ("somevalueinside", header_data.value_);
 }
 
+TEST(HeaderDataConstructorTest, StringMatchSpecifier) {
```
Why aren't some of the tests in this file now not `DEPRECATED_FEATURE_TEST`?

yangminzhu replied:
Updated to use `DEPRECATED_FEATURE_TEST`.

htuch reviewed (commented):
LGTM, just two small things.

htuch commented on test/common/http/header_utility_test.cc:
```diff
@@ -201,7 +201,7 @@ name: test-header
 }
 
-TEST(HeaderDataConstructorTest, ExactMatchSpecifier) {
+TEST(HeaderDataConstructorTest, DEPRECATED_FEATURE_TEST(ExactMatchSpecifier)) {
```
I guess I'm still vaguely curious why this was passing CI (with compile_time_options rejecting deprecated fields)?

yangminzhu replied:
I'm not sure, by looking at the code, `parseHeaderMatcherFromYaml()` calls the `TestUtility::loadFromYaml(yaml, header_matcher)` which seems essentially called the [MessageUtil::loadFromYaml()](https://github.com/envoyproxy/envoy/blob/fa754faf339d9d9180934af710ede614355b48b3/source/common/protobuf/utility.cc#L403-L413) that does not check for the deprecated fields?

htuch commented on test/common/http/header_utility_test.cc:
```diff
@@ -587,7 +590,7 @@ invert_match: true
 }
 
-TEST(MatchHeadersTest, HeaderPrefixMatch) {
+TEST(MatchHeadersTest, DEPRECATED_FEATURE_TEST(HeaderPrefixMatch)) {
```
Do we have sufficient coverage such that when these are removed, the non-`DEPRECATED_FEATURE_TEST` fully cover the alternative `string_match` behaviors?

yangminzhu replied:
we should have sufficient coverage, my last commit only added `DEPRECATED_FEATURE_TEST` to the unit tests that are **directly** testing the deprecated fields (i.e. `exact_match`, `prefix_match`, etc). I also added the test for the `string_match` field.

That's also why I didn't add the tag to `MayMatchOneOrMoreRequestHeader` but instead changed the test to use `string_match` field because it's testing more than the deprecated fields.

htuch replied:
Makes sense, thanks.

htuch reviewed (commented):
LGTM, just some small test questions.
/wait

markdroth commented on api/envoy/config/route/v3/route_components.proto:
```diff
@@ -1898,28 +1902,46 @@ message HeaderMatcher {
+    ];
+
+    // If specified, header match will be performed based on the string match of the header value.
+    type.matcher.v3.StringMatcher string_match = 13;
```
@snowp is copying the `StringMatcher` proto to the xds repo in cncf/xds#8.  It might be a good idea to wait for that to land and then use the new type here, since that will save us the trouble of migrating this later.

yangminzhu replied:
the current StringMatcher is used everywhere in many places, is it better to wait and change all usages at once? It is already a large change for control plane to use the StringMatcher to replace the deprecated fields, just want to make it less complicated by not depending on the extra xds repo for now.

markdroth replied:
I believe that @snowp was going to change the existing code that uses `StringMatcher` to be templated so that it would work with either copy of `StringMatcher`.  Once that's done, I think it should be trivial to use the new type here.

I think this just comes down to a question of timing.  Snow, how soon do you think you can land the `StringMatcher` change, including any necessary templatizing of code in Envoy?

snowp replied:
I'm busy the next two days for company perf season, but I'll have more time to work on it starting Wednesday. Early review of https://github.com/envoyproxy/envoy/pull/17096 could be helpful in speeding this up to make sure that there is agreement on the direction.

yangminzhu commented:
I'm not sure why the following tests only failed in CI but passed in my local run, @htuch  @mattklein123  @qiwzhang do you might have any ideas about this? I checked the jwt filter code but has no idea how is it even affected by this PR. thanks.

```
[ RUN      ] Protocols/RemoteJwksIntegrationTest.WithGoodTokenAsyncFetchFast/IPv4_HttpDownstream_Http2Upstream
[2021-07-09 23:02:04.849][15090][critical][assert] [test/integration/http_integration.cc:470] assert failure: 0. Details: Timed out waiting for new connection.
```

```
[ RUN      ] Protocols/RemoteJwksIntegrationTest.WithGoodTokenAsyncFetchFast/IPv4_HttpDownstream_Http2Upstream
[       OK ] Protocols/RemoteJwksIntegrationTest.WithGoodTokenAsyncFetchFast/IPv4_HttpDownstream_Http2Upstream (68 ms)
[ RUN      ] Protocols/RemoteJwksIntegrationTest.WithGoodTokenAsyncFetchFast/IPv4_Http2Downstream_HttpUpstream
[2021-07-09 23:02:22.138][16531][critical][assert] [test/integration/http_integration.cc:470] assert failure: 0. Details: Timed out waiting for new connection.
```

htuch commented:
@yangminzhu not sure, but with integration tests, machine load and timing can have an impact on flaky tests. Try the advice at https://github.com/envoyproxy/envoy/blob/main/test/integration/README.md#debugging-integration-tests.

yangminzhu commented:
> @yangminzhu not sure, but with integration tests, machine load and timing can have an impact on flaky tests. Try the advice at https://github.com/envoyproxy/envoy/blob/main/test/integration/README.md#debugging-integration-tests.

Thanks for the suggestion. I reproduced the test flaky with `--runs_per_test=1000` in a clean repo at `main` branch which makes me think this PR is unrelated to the test failure. I created https://github.com/envoyproxy/envoy/issues/17303 for tracking the issue.

@mattklein123 @htuch Could you approve the PR if there are no other comments so that we can get this merged while @qiwzhang is working on the test flaky? thanks.

htuch reviewed (approved):
LGTM, thanks!
